### ISSUE TITLE
Race between storing hardlink master and later hardlink references

### ISSUE THREAD

**Issue by enkore** [borgbackup/borg#1297]
https://github.com/borgbackup/borg/blob/1.0-maint/borg/archive.py#L625

- inodes are recycled, so an (dev, inode) pair doesn't really identify a file during a backup run
- If there's an error later on, then `self.hard_links[st.st_ino, st.st_dev] = safe_path` would have created a reference to an item that didn't make it into the archive
- I'm not sure, it may be possible to sneak in a reference to file contents with a stat result that doesn't belong to them
  1. have file with inode x, contents y and permission z
  2. have a hardlink there
  3. get rid of the inode (to delete an inode one may not have the necessary permissions to read the contents)
  4. create new file with permissions you like, lodge it so that they have the same inode, create another dirent for the file so that you get nlink > 1
  5. assuming you manage the proper timing:
     - the file (x, y, z) will end up in the archive as usual
    - the new file (x, y, z') has the same contents (due to being recognized as a hard link), the ones of file (x, y, z), but permissions you can control z'
    - This is kinda low-impact though. Not only requires that a very peculiar permissions layout, but also precise timing and the ability to control/predict inodes (which may be not hard or may be very hard)

**Comment by enkore** [comment]
Close wontfix?

**Comment by enkore** [comment]
A possible fix, if this does turn out to be a relevant issue, would be to use the item metadata (esp. permission bits, ACL) from the hardlink master, either in create code or in all other code.

**Comment by enkore** [comment]
Yet another option would be to also take the inode generation number into account (but only for the hard_links mapping, nowhere else). On Linux the number may be retrieved for some filesystems via the ioctl(2) FS_IOC_GETVERSION (thanks @verygreen).

**Comment by ThomasWaldmann** [comment]
note: the symlink at the very top of this issue points to a changing place, it should have a commit hash or tag rather than "1.0-maint" in it.

**Comment by enkore** [comment]
correct. i think a quick fix would be to just move the assignment down.

**Comment by ThomasWaldmann** [comment]
Yup, sounds good. You do it?
